### PR TITLE
Issue #53

### PR DESCRIPTION
**Pull request by JoseCanavire**
Issue #53  …
* I am adding test cases for the new features added.
It is a big pr, comments are welcomed and I plan to add more tests.
* This is just for android
* Properties need to be adapted to run the mobile tests changing
  * selenium_host
  * uncomment android local mobile properties.
* Add newer version of the apk.

### CODE REVIEW

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/CarinaDescriptionPage.java [hunk 1]
```diff
@@ -17,10 +17,54 @@
...
+
+	@FindBy(className = "android.widget.ImageButton")
+	private ExtendedWebElement leftMenuButton;
+
 	public CarinaDescriptionPage(WebDriver driver) {
 		super(driver);
 	}
 
+	@Override
+	public WebViewBase navitateToWebview() {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/CarinaDescriptionPage.java]: Please correct the method name.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/Charts.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.qaprosoft.carina.demo.mobile.gui.pages.android;
+
+import com.qaprosoft.carina.core.foundation.utils.factory.DeviceType;
+import com.qaprosoft.carina.demo.mobile.gui.pages.common.ChartsBase;
+import org.openqa.selenium.WebDriver;
+
+@DeviceType(pageType = DeviceType.Type.ANDROID_PHONE, parentClass = ChartsBase.class)
+public class Charts extends ChartsBase {
+    public Charts(WebDriver driver) { super(driver); }
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/Charts.java]: Please apply carina formatter for it.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/Maps.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.qaprosoft.carina.demo.mobile.gui.pages.android;
+
+import com.qaprosoft.carina.core.foundation.utils.factory.DeviceType;
+import com.qaprosoft.carina.demo.mobile.gui.pages.common.MapsBase;
+import org.openqa.selenium.WebDriver;
+
+@DeviceType(pageType = DeviceType.Type.ANDROID_PHONE, parentClass = MapsBase.class)
+public class Maps extends MapsBase {
+    public Maps(WebDriver driver) { super(driver); }
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/Maps.java]: please apply carina formatter

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/CarinaDescriptionPage.java [hunk 4]
```diff
@@ -17,10 +17,54 @@
...
+
+	@FindBy(xpath = "//android.widget.CheckedTextView[@text = 'Charts']")
+	private ExtendedWebElement chartsLink;
+
+	@FindBy(xpath = "//android.widget.CheckedTextView[@text = 'Map']")
+	private ExtendedWebElement mapLink;
+
+	@FindBy(xpath = "//android.widget.CheckedTextView[@text = 'UI elements']")
+	private ExtendedWebElement uiElementsLink;
+
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/CarinaDescriptionPage.java]: please apply carina formatter.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/ChartsBase.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.qaprosoft.carina.demo.mobile.gui.pages.common;
+
+import com.qaprosoft.carina.core.gui.AbstractPage;
+import org.openqa.selenium.WebDriver;
+
+public class ChartsBase extends AbstractPage {
+    public ChartsBase(WebDriver driver) { super(driver); }
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/ChartsBase.java]: please apply carina formatter

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/MapsBase.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.qaprosoft.carina.demo.mobile.gui.pages.common;
+
+import com.qaprosoft.carina.core.gui.AbstractPage;
+import org.openqa.selenium.WebDriver;
+
+public class MapsBase extends AbstractPage {
+    public MapsBase(WebDriver driver) { super(driver); }
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/MapsBase.java]: Please apply carina formatter

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java [hunk 7]
```diff
@@ -1,13 +1,20 @@
...
 import com.qaprosoft.carina.core.gui.AbstractPage;
+import org.openqa.selenium.WebDriver;
 
 public abstract class CarinaDescriptionPageBase extends AbstractPage {
 
 	public CarinaDescriptionPageBase(WebDriver driver) {
 		super(driver);
 	}
 
+	public abstract WebViewBase navitateToWebview();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java]: - please correct navitateToWebview
- is WebViewBase Page or not?

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.events.EventFiringWebDriver;
+
+import static io.appium.java_client.touch.offset.PointOption.point;
+
+@DeviceType(pageType = DeviceType.Type.ANDROID_PHONE, parentClass = UIElementsBase.class)
+public class UIElements extends UIElementsBase {
+    @FindBy(id = "editText")
+    private ExtendedWebElement text;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: please rename it from private ExtendedWebElement text to private ExtendedWebElement textField

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+    @FindBy(id = "editText2")
+    private ExtendedWebElement email;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: same as described above.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    @FindBy(id = "editText3")
+    private ExtendedWebElement date;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: same as described above.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    private ExtendedWebElement maleRadioButton;
+
+    @FindBy(id = "radioButton3")
+    private ExtendedWebElement femaleRadioButton;
+
+    @FindBy(id = "radioButton5")
+    private ExtendedWebElement otherRadioButton;
+
+    @FindBy(id = "radioButton5")
+    private ExtendedWebElement seekBar;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: please rename it from seekBar to seekBarButton

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+
+    public void fillInText(String text) {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: please rename it 
from fillInText 
to typeText

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+        this.text.type(text);
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: There is no sense to use "this" here

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public void fillInEmail(String email) {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - Please rename it from "fillInText" to "typeTextFld"
- There is no sense to use "this" here

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+        this.email.type(email);
+    }
+
+    public void fillInDate(String date) {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - Same comment as described above

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        this.date.type(date);
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - There is no sense to use "this" here

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public void selectMale() {
+        this.maleRadioButton.click();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - There is no sense to use "this" here

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public void selectFemale() {
+        this.femaleRadioButton.click();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - There is no sense to use "this" here

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    public void selectOther() {
+        this.otherRadioButton.click();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - There is no sense to use "this" here

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public void checkCopy() {
+        ScrollToElement(checkBox);
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - Please rename "ScrollToElement" according to Java Naming Conventions

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        this.checkBox.click();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - There is no sense to use "this" here

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+
+    public UIElements(WebDriver driver) {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - Is "UIElements" page or screen? 
If "Screen" then UIElementsScreen 
If "Page" then UIElementsPage

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+        super(driver);
+    }
+
+    private AndroidElement ScrollToElement(ExtendedWebElement element) {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - Please rename "ScrollToElement" according to Java Naming Conventions

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+import com.qaprosoft.carina.demo.mobile.gui.pages.common.WebViewBase;
+import io.appium.java_client.AppiumDriver;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.events.EventFiringWebDriver;
+
+import java.util.Set;
+
+@DeviceType(pageType = DeviceType.Type.ANDROID_PHONE, parentClass = WebViewBase.class)
+public class WebView extends WebViewBase {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java]: - Please rename "WebView" according to Project Naming Conventions

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+        webViewDriver.findElement(BY_EMAIL).sendKeys("contact@email.com");
+        webViewDriver.findElement(BY_NAME).sendKeys("John Doe");
+        webViewDriver.findElement(BY_QUESTION).sendKeys("Your framework is awesome");
+        webViewDriver.hideKeyboard();
+        webViewDriver.findElement(BY_SUBMIT).click();
+        webViewDriver.context(NATIVE_APP);
+    }
+
+    public boolean wasSuccessfulMessageSent(){
+        boolean found= false;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java]: - Pleas apply carina formatter

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java [hunk 26]
```diff
@@ -1,13 +1,20 @@
...
+
+	public abstract ChartsBase navigateToCharts();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java]: - Same comment as described above.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java [hunk 27]
```diff
@@ -1,13 +1,20 @@
...
+
+	public abstract MapsBase navigateToMap();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java]: - Same comment as described above.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java [hunk 28]
```diff
@@ -1,13 +1,20 @@
...
+
+	public abstract UIElementsBase navigateToUIElements();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/CarinaDescriptionPageBase.java]: - Same comment as described above.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/WebViewBase.java [hunk 29]
```diff
@@ -0,0 +1,19 @@
+package com.qaprosoft.carina.demo.mobile.gui.pages.common;
+
+import com.qaprosoft.carina.core.gui.AbstractPage;
+import org.openqa.selenium.WebDriver;
+
+public abstract class WebViewBase extends AbstractPage {
+
+    public WebViewBase(WebDriver driver) {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/WebViewBase.java]: - Please apply carina formatter

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java [hunk 30]
```diff
@@ -20,6 +20,26 @@ public CarinaDescriptionPage(WebDriver driver) {
...
 	}
 
+	@Override
+	public WebViewBase navitateToWebview() {
+		return null;
+	}
+
+	@Override
+	public ChartsBase navigateToCharts() {
+		return null;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java]: - Please use instead of "return null" something like  "throw new UnsupportedOperationException(METHOD_ISN_T_IMPLEMENTED_FOR_ANDROID_MESSAGE);"

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java [hunk 31]
```diff
@@ -20,6 +20,26 @@ public CarinaDescriptionPage(WebDriver driver) {
...
+	}
+
+	@Override
+	public MapsBase navigateToMap() {
+		return null;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java]: - Same as described above.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java [hunk 32]
```diff
@@ -20,6 +20,26 @@ public CarinaDescriptionPage(WebDriver driver) {
...
+	}
+
+	@Override
+	public UIElementsBase navigateToUIElements() {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java]: - Same as describe above.
- Please rename according to Project Naming Convenience.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.events.EventFiringWebDriver;
+
+import java.util.Set;
+
+@DeviceType(pageType = DeviceType.Type.ANDROID_PHONE, parentClass = WebViewBase.class)
+public class WebView extends WebViewBase {
+
+    private AppiumDriver webViewDriver;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java]: - Please rewrite whole WebViewPage according to Project style guide.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/WebViewBase.java [hunk 34]
```diff
@@ -0,0 +1,19 @@
...
+        super(driver);
+    }
+
+    public abstract void goToContactUsPage();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/WebViewBase.java]: Why goToContactUsPage is abstract?

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/UIElementsBase.java [hunk 35]
```diff
@@ -0,0 +1,23 @@
...
+
+    public abstract void fillInDate(String date);
+
+    public abstract void selectMale();
+
+    public abstract void selectFemale();
+
+    public abstract void selectOther();
+
+    public abstract void checkCopy();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/UIElementsBase.java]: - Please explain why checkCopy is abstract?

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 36]
```diff
@@ -0,0 +1,105 @@
...
+
+    public void fillInEmail(String email) {
+        this.email.type(email);
+    }
+
+    public void fillInDate(String date) {
+        this.date.type(date);
+    }
+
+    public void selectMale() {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - Please rename it according Project Naming Convenience.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java [hunk 37]
```diff
@@ -0,0 +1,105 @@
...
+        this.maleRadioButton.click();
+    }
+
+    public void selectFemale() {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: - Same comment as described above.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java [hunk 38]
```diff
@@ -20,6 +20,26 @@ public CarinaDescriptionPage(WebDriver driver) {
 		super(driver);
 	}
 
+	@Override
+	public WebViewBase navitateToWebview() {
+		return null;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/ios/CarinaDescriptionPage.java]: Please use instead of "return null" something like "throw new UnsupportedOperationException(METHOD_ISN_T_IMPLEMENTED_FOR_ANDROID_MESSAGE);"

File: src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java [hunk 39]
```diff
@@ -34,6 +33,36 @@ public void testLoginUser() {
 		CarinaDescriptionPageBase carinaDescriptionPage = loginPage.clickLoginBtn();
 		Assert.assertTrue(carinaDescriptionPage.isPageOpened(), "Carina description page isn't opened");
 	}
+
```
Comment (ABialkevich) [inline src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java]: - Please rewrite tests according to updated pages and comments.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java (hunk 23 above)
Comment (JoseCanavire) [reply src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElements.java]: swipe is not working for android here. There is a carrusel in the first half of the screen. So, the scrolling as is doesn't work either.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java (hunk 33 above)
Comment (JoseCanavire) [reply src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebView.java]: Do you have an example where you had to switch between contexts and use the pageObject pattern?

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElementsPage.java [hunk 40]
```diff
@@ -0,0 +1,105 @@
...
+    private ExtendedWebElement maleRadioButton;
+
+    @FindBy(id = "radioButton3")
+    private ExtendedWebElement femaleRadioButton;
+
+    @FindBy(id = "radioButton5")
+    private ExtendedWebElement otherRadioButton;
+
+    @FindBy(id = "radioButton5")
+    private ExtendedWebElement seekBarButton;
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElementsPage.java]: - Rename from seekBarButton to seekBarRadioButton

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElementsPage.java [hunk 41]
```diff
@@ -0,0 +1,105 @@
...
+
+    public void typeDate(String date) {
+        dateField.type(date);
+    }
+
+    public void selectMale() {
+        maleRadioButton.click();
+    }
+
+    public void selectFemale() {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElementsPage.java]: - Please rename from selectFemale to clickFemaleRadioButton

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElementsPage.java [hunk 42]
```diff
@@ -0,0 +1,105 @@
...
+
+    public void typeEmail(String email) {
+        emailField.type(email);
+    }
+
+    public void typeDate(String date) {
+        dateField.type(date);
+    }
+
+    public void selectMale() {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/UIElementsPage.java]: - Please rename method as 61 row

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/CarinaDescriptionPage.java [hunk 43]
```diff
@@ -1,29 +1,72 @@
...
+    private ExtendedWebElement leftMenuButton;
+
+    public CarinaDescriptionPage(WebDriver driver) {
+        super(driver);
+    }
+
+    @Override
+    public WebViewPageBase navigateToWebviewPage() {
+        leftMenuButton.click();
+        webviewLink.click();
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/CarinaDescriptionPage.java]: - Please webviewLink rename to webViewLink

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/WebViewPageBase.java [hunk 44]
```diff
@@ -0,0 +1,19 @@
...
+    public WebViewPageBase(WebDriver driver) {
+        super(driver);
+    }
+
+    public abstract void goToContactUsPage();
+
+    public abstract void fillContactPage();
+
+    public abstract boolean isSuccessfullMessagePresent();
+
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/WebViewPageBase.java]: - Please remove extra row.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/MapsPageBase.java [hunk 45]
```diff
@@ -0,0 +1,9 @@
+package com.qaprosoft.carina.demo.mobile.gui.pages.common;
+
+import com.qaprosoft.carina.core.gui.AbstractPage;
+import org.openqa.selenium.WebDriver;
+
+public class MapsPageBase extends AbstractPage {
+    public MapsPageBase(WebDriver driver) { super(driver); }
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/MapsPageBase.java]: - Please apply PR carina formatter.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java [hunk 46]
```diff
@@ -0,0 +1,77 @@
...
+    private ExtendedWebElement questionField;
+
+    @FindBy(xpath = "//button[@type='submit' and text()='Send']")
+    private ExtendedWebElement submitButton;
+
+    @FindBy(xpath = "//div[contains(@class,'t-form__successbox')]")
+    private ExtendedWebElement successLabel;
+
+    public WebViewPage(WebDriver driver) {
+        super(driver);
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java]: - Please rewrite constructor according to project style guide.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java [hunk 47]
```diff
@@ -0,0 +1,29 @@
...
+    @FindBy(className = "t708__btn")
+    private ExtendedWebElement contactUsLink;
+
+    public WebViewPage(WebDriver driver) {
+        super(driver);
+    }
+
+    @Override
+    public ContactUsPageBase goToContactUsPage() {
+        changeToWebViewContext(getDriver());
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java]: - Please remove changeToWebViewContext from method and add if you need into test

File: src/main/resources/_config.properties [hunk 48]
```diff
@@ -52,7 +52,7 @@ thread_count=1
 
 #================ Report configuration ===============#
 core_log_level=INFO
-auto_screenshot=true
+auto_screenshot=false
```
Comment (ABialkevich) [inline src/main/resources/_config.properties]: please revert this changes

File: src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java [hunk 49]
```diff
@@ -1,50 +1,88 @@
...
+        CarinaDescriptionPageBase carinaDescriptionPage = loginPage.login();
+        UIElementsPageBase uiElements = carinaDescriptionPage.navigateToUIElementsPage();
+        uiElements.typeText("some Text");
+        uiElements.typeDate("22/10/2018");
+        uiElements.typeEmail("some@email.com");
+        uiElements.swipeToFemaleRadioButton();
+        uiElements.checkCopy();
+        uiElements.clickOnFemaleRadioButton();
+        uiElements.clickOnOtherRadioButton();
+
```
Comment (ABialkevich) [inline src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java]: It looks strange without verification. Is it expected?

File: src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java [hunk 50]
```diff
@@ -1,50 +1,88 @@
...
+    }
+
+    private void setApplicationPath() {
```
Comment (ABialkevich) [inline src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java]: Please use config.properties for app settings.

File: src/main/java/utils/MobileContextUtils.java [hunk 51]
```diff
@@ -0,0 +1,85 @@
...
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Proxy;
+import java.util.Set;
+
+public class MobileContextUtils {
+
+    private static final Logger LOGGER = Logger.getLogger(DriverHelper.class);
+
+    private static AppiumDriver getDriverSafe(WebDriver driver) {
```
Comment (ABialkevich) [inline src/main/java/utils/MobileContextUtils.java]: There is will be problem with multithreading

Comment (ABialkevich) [reply src/main/java/utils/MobileContextUtils.java]: I believe it would be much more better.
public WebDriver getDriverSafe() {
        WebDriver driver = getDriver();
        if (driver instanceof EventFiringWebDriver) {
            driver = ((EventFiringWebDriver) drv).getWrappedDriver();
        }
        return driver;
    }

File: src/main/java/utils/MobileContextUtils.java [hunk 52]
```diff
@@ -0,0 +1,85 @@
...
+                }
+                if (appiumDriver instanceof AppiumDriver) {
+                    return (AppiumDriver) appiumDriver;
+                }
+            }
+        }
+        throw new ClassCastException("Appium Driver can not be casted from the actual driver.");
+    }
+
+    public void switchMobileContext(View context, WebDriver dvr) {
```
Comment (ABialkevich) [inline src/main/java/utils/MobileContextUtils.java]: There is no sense to pass Driver into method.

File: src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java [hunk 53]
```diff
@@ -1,50 +1,85 @@
...
+
+    @Test(description = "JIRA#DEMO-0011")
+    @MethodOwner(owner = "qpsdemo")
+    public void testWebView() {
+        WelcomePageBase welcomePage = initPage(getDriver(), WelcomePageBase.class);
+        LoginPageBase loginPage = welcomePage.clickNextBtn();
+        loginPage.login();
+        WebViewPageBase webViewPageBase = initPage(getDriver(), WebViewPageBase.class);
+        MobileContextUtils contextHelper = new MobileContextUtils();
+        contextHelper.switchMobileContext(MobileContextUtils.View.WEB, getDriver());
```
Comment (ABialkevich) [inline src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java]: contextHelper.switchMobileContext(MobileContextUtils.View.WEB, getDriver()); it looks strange.
I believe it should be like that:
contextHelper.switchMobileContext(View.WEB);

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java [hunk 54]
```diff
@@ -0,0 +1,28 @@
...
+import com.qaprosoft.carina.core.foundation.webdriver.decorator.ExtendedWebElement;
+import com.qaprosoft.carina.demo.mobile.gui.pages.common.ContactUsPageBase;
+import com.qaprosoft.carina.demo.mobile.gui.pages.common.WebViewPageBase;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.support.FindBy;
+
+@DeviceType(pageType = DeviceType.Type.ANDROID_PHONE, parentClass = WebViewPageBase.class)
+public class WebViewPage extends WebViewPageBase {
+
+
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java]: - please remove extra row

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java [hunk 55]
```diff
@@ -0,0 +1,28 @@
...
+    public WebViewPage(WebDriver driver) {
+        super(driver);
+    }
+
+    @Override
+    public ContactUsPageBase goToContactUsPage() {
+        contactUsLink.click();
+        return initPage(getDriver(), ContactUsPageBase.class);
+    }
+
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/android/WebViewPage.java]: Please remove extra row.

File: src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/ChartsPageBase.java [hunk 56]
```diff
@@ -0,0 +1,10 @@
+package com.qaprosoft.carina.demo.mobile.gui.pages.common;
+
+import com.qaprosoft.carina.core.gui.AbstractPage;
+import org.openqa.selenium.WebDriver;
+
+public class ChartsPageBase extends AbstractPage {
+    public ChartsPageBase(WebDriver driver) {
```
Comment (ABialkevich) [inline src/main/java/com/qaprosoft/carina/demo/mobile/gui/pages/common/ChartsPageBase.java]: - Please use one code style in PR

File: src/main/java/utils/MobileContextUtils.java [hunk 57]
```diff
@@ -0,0 +1,64 @@
...
+
+public class MobileContextUtils {
+
+    private static final Logger LOGGER = Logger.getLogger(DriverHelper.class);
+
+    private AppiumDriver getDriverSafe() {
+        WebDriver driver = getDriver();
+        if (driver instanceof EventFiringWebDriver) {
+            driver = ((EventFiringWebDriver) driver).getWrappedDriver();
+            if (driver instanceof AppiumDriver) {
```
Comment (ABialkevich) [inline src/main/java/utils/MobileContextUtils.java]: There is no sense in if case.

File: src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java [hunk 58]
```diff
@@ -1,50 +1,86 @@
...
-		if (Type.ANDROID_PHONE.getFamily().equalsIgnoreCase(R.CONFIG.get("capabilities.platformName")))
-			pathToApp = androidPath;
-		else 
-			pathToApp = iosPath;
-		File file = new File(pathToApp);
-		Configuration.setMobileApp(file.getAbsolutePath());
-	}
+    @BeforeClass
+    public void setup() {
+        String pathToApp = R.CONFIG.get("application.path");
```
Comment (ABialkevich) [inline src/test/java/com/qaprosoft/carina/demo/MobileSampleTest.java]: I don't see application.path in config.properties. Where do you get it?
